### ISSUE TITLE
Interpolated DOAS data when exact data point exists

### ISSUE THREAD

**Issue by twVolc** [twVolc/PyCamPermanent#209]
I'm suddenly seeing a lot of `Interpolated DOAS data for image {}` in RTP (see top screenshot), even though the data for these specific times are available (see bottom screenshot). I'm not 100% certain but I think this has only started appearing since #206. Could this be because data aren't being downloaded in order or another problem? I guess because we are catching the `BaseException` in `update_doas_calibration()` it's hard to know exactly what is going on. For some reason we aren't finding the correct DOAS data point, but seem to find some before and after data points to interpolate. 

It's also odd that it isn't throwing the  `'No DOAS data point within {}s of image time: {}. '` notification, as all DOAS data points other than the correct time will be over the 2s interval used as a limit for interpolation. This must mean the code is finding DOAS data point for the correct time for that image pair, so why are we ending up in that `BaseException` in the first place? 

I'm guessing the #206 code update might raise an exception itself? Perhaps that while-loop doesn't execute properly? 
``` while (datetime.datetime.now() < timeout) and not np.any(img_time < self.doas_worker.results.index):```
@ubdbra001 were you able to check execution of this block? Other than this I can't see why the code has suddenly started going to the BaseException when presumably the DOAS data point is available.

It's probable that we're ending up with the same performance as if we just correctly extracted the DOAS point in the first place, since we will be interpolating to an exact time that exists, but we're executing much more code than we should and it's not the correct execution.

![image](https://github.com/user-attachments/assets/28b84d8c-3417-484a-9213-1a7d86879efd)
![image](https://github.com/user-attachments/assets/076faa22-983b-46e6-8493-ae05198d44b9)

**Comment by twVolc** [comment]
It's not an error with the execution of the while-loop specifically - I've added a print statement to investigate the exception that is being caught it's printing this:
![image](https://github.com/user-attachments/assets/b2b43ba1-85c6-4fd2-a519-228ff4ec6269)
So I'm a bit confused what's happening - it seems the loop is failing to find the `img_time` for a specific spectrum. But it must then be finding it for it to perform the interpolation...

Probably need to investigate what's going on inside the while loop - why isn't this returning a column density:
```
cd = self.doas_worker.results.get(img_time)
if cd is not None: break
time.sleep(0.5)
```
But it seems it used to in the old code.

**Comment by twVolc** [comment]
NOTE: This doesn't seem to happen on every data point - some `processed_pair` prints in the terminal don't have any associated issue with finding a spectrum.

**Comment by twVolc** [comment]
I've removed `and not np.any(img_time < self.doas_worker.results.index):` from the while-loop statement and I'm not seeing this message appear at all any more. This might suggest that spectra are arriving (or being processed) out of order, but I can't see how this is the case as this still wouldn't explain why within the exception we aren't getting the `'No DOAS data point within {}s of image time: {}.` message. 

I'm wondering if perhaps the `and not np.any(img_time < self.doas_worker.results.index):` isn't doing the exact check we're wanting it to, so the while-loop exits immediately and goas to `else`? I can't think of another explanation, but as far as I understand that section of code it does seem to do what we want it to do. 

I'm a bit stumped...

**Comment by twVolc** [comment]
Just a note, this isn't exclusive to RTP, so definitely isn't to do with spectra arriving at different times. Post-processing exhibits the same behaviour. This should at least make it quicker and easier to debug.

**Comment by ubdbra001** [comment]
> Just a note, this isn't exclusive to RTP, so definitely isn't to do with spectra arriving at different times. Post-processing exhibits the same behaviour. This should at least make it quicker and easier to debug.

Okay, that does make it easier. I did check it and didn't notice anything like this, but it wasn't for too long. 
I'll dig in and see what's going on.

**Comment by ubdbra001** [comment]
Okay, the reason it does this is post-processing is that it checks to see if there are later results before checking for the actual initial time. So the loop will never start and throw the `KeyError` for all the data points (except the last one, probably). 
I thought I was being to clever with this approach. I'll move the check for later data points after `if cd is not None: break` which should fix it for post processing, the see what's going on with RTP.

**Comment by twVolc** [comment]
Ah yes, good spot! My guess would be exactly the same thing is happening in RTP - spectra processing is quite fast compared to image processing, so it's probably reasonably likely we process a later spectra before `process_pair()` gets to this part of the code. That would explain why we somtimes see it and sometimes don't.

**Comment by twVolc** [comment]
> `if cd is not None: break`

I guess you can add it specifically to this statement as an `or`?

**Comment by ubdbra001** [comment]
> Thinking about it more - when a spectrum doesn't exist for a specfic time, this part of code hangs because of the timeout. That's quite a long timeout in which the spectrum processing could be continuing, so once this code finally breaks out of the timeout there will be an accumulation of spectra that can quickly be processed and added to `results` before the next image pair gets back to this part of the code.

That's a good point, and I'd put good money on that being the issue. So, as you say, checking for later points should come after checking for the `img_time` so at least it has a chance of seeing if the data point has shown up before checking for later ones and exiting.

**Comment by ubdbra001** [comment]
> I guess you can add it specifically to this statement as an `or`?

No, if you use break then it doesn't proceed to the `else` condition, so it wouldn't raise the error and would skip everything within the `except` part. It's currently a seperate if-statement after `if cd is not None: break` but could be an `elif` there.

**Comment by twVolc** [comment]
> No, if you use break then it doesn't proceed to the `else` part of the while-loop, so it wouldn't raise the error and would skip everything within the `except` part. It's currently a seperate if-statement after `if cd is not None: break` but could be an `elif` there.

Yeh sorry I realised this afterwards, deleted the comment haha

**Comment by ubdbra001** [comment]
Hmmm, initial runs seem to have it waiting for the entire timeout period. It doesn't show newer data points even though newer ones have come in.

Given you know more about threading and locking than me (especially with regards this code), do you think the changes you made for #205 stop the doas results being updated while it's waiting?

**Comment by twVolc** [comment]
> Hmmm, initial runs in RTP seem to have it waiting for the entire timeout period. It doesn't show newer data points even though newer ones have come in.
> 
> Given you know more about threading and locking than me (especially with regards this code), do you think the changes you made for #205 stop the doas results being updated while it's waiting?

Sorry yes, I actually started writing this and then decided not to bombard you with information, but I guess this is actually very important here.

My update has put a lock on threads in place - `line 2296` in `so2_camera_processor.py` acquires this lock, which means any other thread that tries to acquire the lock won't be able to until the `PyplisWorker` thread lets go of it. `IFitWorker.add_doas_results()` tries to acquire the lock, as this is the part that edits `IFitWorker.results` so is the part of this thread that we don't want running if it might edit that attribute as `PyplisWorker` is reading the attribute - I think it's possible this was causing the issue in the first place. This lock makes the code thread safe, but I guess it does mean that `IFitWorker` won't be able to update results until after `PyplisWorker` is completely done with it. Even with this lock, I still think it's possible that `IFitWorker` could every now and then add 2 spectra before `PyplisWorker.process_pair()` has managed to process the first image, simply because of the speed at which the spectra are processed - so this is still probably the reason we hit the while-loop error in RTP. But I guess this lock solution brings its own issue in not allowing us to progress faster than the timeout in most situations. I'm not sure how best to deal with this. One solution might be to only acquire the lock at certain points in `PyplisWorker.update_doas_calibration()`, rather than acquiring it for the entirety of that function. But it would probably need a bit of careful thinking to determine where that lock needs to be acquired so that it still performs its function of preventing race conditions, but doesn't hold things for the entirety of the timeout period.

Sorry, that's a long explanation with no clear solution yet. Does this make sense though?

**Comment by ubdbra001** [comment]
Yes, it does and explains exactly what's going on here. Will need to have a think about how to resolve it though.
In the meantime I'll revert the additional condition to the while-loop, and add the ability to change the timeout in the config. That way at least it'll be easier to adjust it.

**Comment by twVolc** [comment]
Great. I think there will still be times where your check after `if cd is not None: break` will still execute and speed things up by bypassing the timeout, but like you mention, it won't happen all the time so isn't ideal. It might still be worth including as an initial step though as we then just need to think about how we incorporate/edit the locking around that.

**Comment by ubdbra001** [comment]
Given the locking, do you think that the result associated with a time point will ever be found if it isn't already there when `update_doas_calibration()` is called and access is locked?
My gut feeling is 'no' as the results array shouldn't be able to update until `update_doas_calibration()` is finished.

**Comment by ubdbra001** [comment]
I've added the changes to [this branch](https://github.com/twVolc/PyCamPermanent/tree/issue-209/fix-the-fix), but I'm increasingly convinced that the lock will prevent the doas result from being updated, and so if the timepoint is not there it will just timeout regardless.

**Comment by ubdbra001** [comment]
Maybe fixed by #217 waiting to hear back from Indonesia team
